tests/python/pants_test/engine/test_rules.py

```diff
@@ -244,18 +246,22 @@ def d_from_a(a: A) -> D:
     with self.assertRaises(Exception) as cm:
       create_scheduler(rules)
 
-    self.assert_equal_with_printing(dedent("""
-                     Rules with errors: 3
-                       (A, [B, C], a_from_b_and_c()):
-                         Was not usable by any other @rule.
-                       (A, [C, B], a_from_c_and_b()):
-                         Was not usable by any other @rule.
-                       (D, [A], d_from_a()):
-                         Ambiguous rules to compute A with parameter types (B+C):
-                           (A, [B, C], a_from_b_and_c()) for (B+C)
-                           (A, [C, B], a_from_c_and_b()) for (B+C)
-                     """).strip(),
-      str(cm.exception))
+    self.assert_equal_with_printing(
+      dedent(
+        """\
+        Rules with errors: 3
+          Rule(func=a_from_b_and_c(), product=A, params=[B, C]):
+            Was not reachable, possibly because it was shadowed by another rule.
```

Comment:
The first two of these errors aren't the useful ones; the third one is the one which really points at the core problem. Ideally this would either _just_ show the third one, or at least, show the third one first... Is either of those easy-ish to do?